src/infrastructure/channels/reliable_channel.rs

```diff
@@ -0,0 +1,146 @@
+use super::Channel;
+
+use net::{LocalAckRecord, ExternalAcks, NetworkConfig, NetworkQuality, NetworkQualityMeasurer};
+use packet::header::{HeaderParser, HeaderReader, AckedPacketHeader, StandardHeader};
+use sequence_buffer::{SequenceBuffer, CongestionData};
+use infrastructure::{DeliveryMethod, Fragmentation};
+use error::{PacketErrorKind,NetworkResult};
+use packet::{Packet, PacketData,PacketTypeId};
+
+use std::io::{Cursor, Read};
+use std::time::Instant;
+
+/// This channel should be used for reliable processing of packets.
+///
+/// *Details*
+///
+///  1. Reliable.
+///  2. Guarantee of delivery.
+///  3. Option to specify if data needs to be ordered.
+///  4. Packets will not be dropped.
+///  5. Duplication not possible
+pub struct ReliableChannel {
+    // settings
+    ordered: bool,
+    config: NetworkConfig,
+
+    // reliability control
+    seq_num: u16,
+    waiting_packets: LocalAckRecord,
+    their_acks: ExternalAcks,
+    dropped_packets: Vec<Box<[u8]>>,
+
+    // congestion control
+    network_quality_measurer: NetworkQualityMeasurer,
+    congestion_data: SequenceBuffer<CongestionData>,
+    quality: NetworkQuality,
+    rtt: f32,
+}
+
+impl ReliableChannel {
+    /// Create a new instance of the unreliable channel by specifying if channel needs to order incoming packets.
+    pub fn new(ordered: bool, config: NetworkConfig, ) -> ReliableChannel {
+        ReliableChannel {
+            // settings
+            ordered,
+            config: config.clone(),
+
+            // reliability control
+            seq_num: 0,
+            waiting_packets: Default::default(),
+            their_acks: Default::default(),
+            dropped_packets: Vec::new(),
+
+            // congestion control
+            network_quality_measurer: NetworkQualityMeasurer::new(config),
+            congestion_data: SequenceBuffer::with_capacity(<u16>::max_value() as usize),
+            quality: NetworkQuality::Good,
+            rtt: 0.0,
+        }
+    }
+}
+
+impl Channel for ReliableChannel {
+    /// This will prepossess an reliable packet.
+    ///
+    /// 1. Add congestion data entry so that it could be monitored.
+    /// 2. Queueing new packet for acknowledgement system.
+    /// 4. Fragmentation of the payload.
```

Comment:
Maybe add "if needed"? This implies the payload is always fragmented.

Reply:
It kind of does for reliable packets right now. Unreliable packets processed by UnreliableChannel will not be fragmented.